### ISSUE TITLE
Quarkus doesn't detect Docker rootless anymore

### ISSUE THREAD

**Issue by chberger** [quarkusio/quarkus#32246]
### Describe the bug

Starting from ``2.16.4.Final``, Quarkus cannot detect ```Docker rootless``` anymore. Reason for that is a faulty implemenation in ``ContainerRuntimeUtil.getRootlessStateFor``:

```java
private static boolean getRootlessStateFor(ContainerRuntime containerRuntime) {
      ......
            try (InputStream inputStream = rootlessProcess.getInputStream();
                    InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
                    BufferedReader bufferedReader = new BufferedReader(inputStreamReader)) {
                Predicate<String> stringPredicate;
                // Docker includes just "rootless" under SecurityOptions, while podman includes "rootless: <boolean>"
       --->  if (containerRuntime == ContainerRuntime.DOCKER)  {  <-- ContainerRuntime.DOCKER is null 
                    stringPredicate = line -> line.trim().equals("rootless");
                } else {
                    stringPredicate = line -> line.trim().equals("rootless: true");
                }
                return bufferedReader.lines().anyMatch(stringPredicate);
            }
        .....
    }

    /**
     * Supported Container runtimes
     */
    public enum ContainerRuntime {
        DOCKER,
        PODMAN;

        private final boolean rootless;

        ContainerRuntime() {
            this.rootless = getRootlessStateFor(this);
        }

        public String getExecutableName() {
            return this.name().toLowerCase();
        }

        public boolean isRootless() {
            return rootless;
        }
    }
```

This condition ``containerRuntime == ContainerRuntime.DOCKER`` always evaluates to false, since ContainerRuntime.DOCKER is null. That's  because ``getRootlessStateFor`` is being called during constructor execution of ContainerRuntime. Thus we end up in the ``podman`` case, even when ``Docker`` is being detected!

### Expected behavior

Quarkus should detect ``Docker rootless`` properly :-)

### Actual behavior

Quarkus native container builds aren't working on Docker rootless anymore, because the rootless detection fails. Therefor the container isn't started as root, but rather non root which leads to file permission issues ..

### Output of `uname -a` or `ver`

5.4.228 ...... x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

17

### GraalVM version (if different from Java)

GraalVM 22.3.0 Java 17 CE

### Quarkus version or git rev

2.16.4.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.8.6

### Additional information

```
Docker version 20.10.22, build 3a2c30b
runner@....:/$ docker info
Client:
 Context:    default
 Debug Mode: false

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 20.10.22
 Storage Driver: vfs
 Logging Driver: json-file
 Cgroup Driver: none
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 78f51771157abb6c9ed224c22013cdf09962315d
 runc version: v1.1.4-0-g5fd4c4d1
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  **rootless**
 Kernel Version: 5.4.228-132.418.amzn2.x86_64
 Operating System: Ubuntu 22.04.1 LTS (containerized)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.34GiB
 Name: ......
 Docker Root Dir: /home/runner/.local/share/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels: ...
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine

WARNING: Running in rootless-mode without cgroups. To enable cgroups in rootless-mode, you need to boot the system in cgroup v2 mode.
```

**Comment by gsmet** [comment]
I will have a look. I wasn't really happy with this pattern. Looks like I was right :).

**Comment by geoand** [comment]
👍🏼

**Comment by Karm** [comment]
@gsmet Well, this is not a problem in main since we reworked that section to lazy evaluation over here, right?
https://github.com/quarkusio/quarkus/blob/main/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java#L223
...and `this` is not `null` at that time...

**Comment by chberger** [comment]
I did not say that this is null but rather the reference ``ContainerRuntime.DOCKER`` is ``null`` . Because at that time we're still in constructor execution phase ...

**Comment by Karm** [comment]
@chberger I was referencing the current main. Does is fail with the contemporary main / e.g. 3.0.0.CR1 for you too?

**Comment by chberger** [comment]
I haven't tested it with 3.X, my current setup just works with Quarkus 2.X and Docker rootless works only remotely. But you could write an unit test easily. 

Anyway, since you are not calling ``getRootlessStateFor`` in the constructor anymore, I'm pretty sure this will work. When calling ``getRootlessStateFor`` during constructor execution, ``ContainerRuntime.DOCKER`` will always be ``null``. Which seems to be mitigated in 3.X. Anyway, would love to have a patch in ``2.16.X`` because my native container builds are broken. Reason for that is they aren't executed as root anymore, which is crucial to mount the jars etc ...

I know it sounds odd that we need root in Docker rootless to have proper file rights while mounting files from the host, but this is another topic :-)

To solve that I would recommend a quick fix like that:

```
// should work for both, Docker and Podman
line -> line.trim().startsWith("rootless");
```

While doing so, there is no need to distinguish between Podman and Docker. Thus it would work even in 2.16.X again - thanks!

**Comment by Karm** [comment]
@gsmet So this is a case for a backport IMHO

**Comment by gsmet** [comment]
@Karm I decided to simplify things a bit more because the code was a bit hard to follow: https://github.com/quarkusio/quarkus/pull/32309 .

@chberger could you try to build this PR https://github.com/quarkusio/quarkus/pull/32309 locally? You can do it with `./mvnw -Dquickly`. Once done, just execute `./mvnw clean install -f integration-tests/hibernate-validator -Dquarkus.native.container-build=true -Dnative` and see if things work for you.
I would like to make sure we didn't break anything before backporting this series of PRs.
